File: flowcatalyst-go/internal/stream/watcher.go
// Package stream provides MongoDB change stream processing
package stream

import (
	"context"
	"sync"
	"sync/atomic"
	"time"

	"log/slog"
	"go.mongodb.org/mongo-driver/bson"
	"go.mongodb.org/mongo-driver/mongo"
	"go.mongodb.org/mongo-driver/mongo/options"

	"go.flowcatalyst.tech/internal/common/metrics"
)

// StreamConfig holds configuration for a single stream
type StreamConfig struct {
	// Name is the stream name for logging
	Name string

	// SourceCollection is the collection to watch
	SourceCollection string

	// TargetCollection is the collection to write projections to
	TargetCollection string

	// WatchOperations are the operation types to watch (insert, update, replace)
	WatchOperations []string

	// BatchMaxSize is the maximum batch size before flushing
	BatchMaxSize int

	// BatchMaxWait is the maximum time to wait before flushing a batch
	BatchMaxWait time.Duration

	// CheckpointKey is the key for storing checkpoints
	CheckpointKey string
}

// ProjectionMapper maps source documents to projection documents
type ProjectionMapper interface {
	Map(doc bson.M) bson.M
}

// CheckpointStore stores and retrieves resume tokens
type CheckpointStore interface {
	// GetCheckpoint retrieves the resume token for the given key.
	// Returns nil if no checkpoint exists.
	GetCheckpoint(key string) (bson.Raw, error)
	// SaveCheckpoint saves the resume token for the given key.
	SaveCheckpoint(key string, token bson.Raw) error
}

// Watcher watches a MongoDB change stream
type Watcher struct {
	name              string
	client            *mongo.Client
	database          string
	config            *StreamConfig
	checkpointStore   CheckpointStore
	mapper            ProjectionMapper
	targetCollection  *mongo.Collection

	running    bool
	runningMu  sync.Mutex
	ctx        context.Context
	cancel     context.CancelFunc
	wg         sync.WaitGroup

	// Health metrics (like Java's StreamContext)
	batchSequence    atomic.Int64 // Total batches processed
	checkpointedSeq  atomic.Int64 // Last checkpointed sequence
	inFlightCount    atomic.Int32 // Batches currently processing
	fatalError       atomic.Value // Stores error or nil
	availableSlots   atomic.Int32 // Concurrency slots available (default 1 for single-threaded)
}

// NewWatcher creates a new stream watcher
func NewWatcher(
	client *mongo.Client,
	database string,
	config *StreamConfig,
	checkpointStore CheckpointStore,
	mapper ProjectionMapper,
) *Watcher {
	ctx, cancel := context.WithCancel(context.Background())

	w := &Watcher{
		name:             config.Name,
		client:           client,
		database:         database,
		config:           config,
		checkpointStore:  checkpointStore,
		mapper:           mapper,
		targetCollection: client.Database(database).Collection(config.TargetCollection),
		ctx:              ctx,
		cancel:           cancel,
	}

	// Initialize available slots (single-threaded by default)
	w.availableSlots.Store(1)

	return w
}

// Start starts watching the change stream
func (w *Watcher) Start() {
	w.runningMu.Lock()
	if w.running {
		w.runningMu.Unlock()
		slog.Warn("Watcher already running", "stream", w.name)
		return
	}
	w.running = true
	w.runningMu.Unlock()

	w.wg.Add(1)
	go w.watchLoop()

	slog.Info("Stream watcher started", "stream", w.name)
}

// Stop stops the watcher
func (w *Watcher) Stop() {
	slog.Info("Stopping stream watcher", "stream", w.name)

	w.runningMu.Lock()
	w.running = false
	w.runningMu.Unlock()

	w.cancel()
	w.wg.Wait()

	slog.Info("Stream watcher stopped", "stream", w.name)
}

// IsRunning returns true if the watcher is running
func (w *Watcher) IsRunning() bool {
	w.runningMu.Lock()
	defer w.runningMu.Unlock()
	return w.running
}

// GetCurrentBatchSequence returns the total number of batches processed
func (w *Watcher) GetCurrentBatchSequence() int64 {
	return w.batchSequence.Load()
}

// GetLastCheckpointedSequence returns the last checkpointed batch sequence
func (w *Watcher) GetLastCheckpointedSequence() int64 {
	return w.checkpointedSeq.Load()
}

// GetInFlightBatchCount returns the number of batches currently being processed
func (w *Watcher) GetInFlightBatchCount() int32 {
	return w.inFlightCount.Load()
}

// GetAvailableConcurrencySlots returns the available concurrency slots
func (w *Watcher) GetAvailableConcurrencySlots() int32 {
	return w.availableSlots.Load()
}

// HasFatalError returns true if the watcher has encountered a fatal error
func (w *Watcher) HasFatalError() bool {
	return w.fatalError.Load() != nil
}

// GetFatalError returns the fatal error if one occurred
func (w *Watcher) GetFatalError() error {
	if err := w.fatalError.Load(); err != nil {
		return err.(error)
	}
	return nil
}

// setFatalError sets the fatal error and stops the watcher
func (w *Watcher) setFatalError(err error) {
	w.fatalError.Store(err)
	w.runningMu.Lock()
	w.running = false
	w.runningMu.Unlock()
}

// Reconnection settings
const (
	initialBackoff = 5 * time.Second
	maxBackoff     = 60 * time.Second
	backoffMultiplier = 2.0
)

// watchLoop is the main watch loop with automatic reconnection
func (w *Watcher) watchLoop() {
	defer w.wg.Done()
	defer func() {
		w.runningMu.Lock()
		w.running = false
		w.runningMu.Unlock()
	}()

	sourceCollection := w.client.Database(w.database).Collection(w.config.SourceCollection)

	// Build pipeline to filter by operation types
	pipeline := mongo.Pipeline{
		{{Key: "$match", Value: bson.M{
			"operationType": bson.M{"$in": w.config.WatchOperations},
		}}},
	}

	consecutiveFailures := 0
	backoff := initialBackoff

	// Outer loop for reconnection
	for {
		// Check if we should stop
		select {
		case <-w.ctx.Done():
			return
		default:
		}

		// Configure change stream options
		opts := options.ChangeStream().
			SetFullDocument(options.UpdateLookup).
			SetBatchSize(int32(w.config.BatchMaxSize))

		// Resume from checkpoint if available
		if w.checkpointStore != nil {
			if resumeToken, err := w.checkpointStore.GetCheckpoint(w.config.CheckpointKey); err == nil && resumeToken != nil {
				opts.SetResumeAfter(resumeToken)
				slog.Info("Resuming from checkpoint", "stream", w.name)
			} else if err != nil {
				slog.Warn("Failed to load checkpoint, starting from current position", "error", err, "stream", w.name)
			} else {
				slog.Info("No checkpoint found, starting from current position", "stream", w.name)
			}
		}

		slog.Info("Opening change stream",
			"stream", w.name,
			"source", w.config.SourceCollection,
			"target", w.config.TargetCollection,
			"operations", w.config.WatchOperations)

		// Open change stream
		stream, err := sourceCollection.Watch(w.ctx, pipeline, opts)
		if err != nil {
			consecutiveFailures++
			slog.Error("Failed to open change stream, will retry",
				"error", err,
				"stream", w.name,
				"attempt", consecutiveFailures,
				"backoff", backoff)

			// Wait before retry
			select {
			case <-w.ctx.Done():
				return
			case <-time.After(backoff):
			}

			// Increase backoff for next attempt
			backoff = time.Duration(float64(backoff) * backoffMultiplier)
			if backoff > maxBackoff {
				backoff = maxBackoff
			}
			continue
		}

		// Reset backoff on successful connection
		consecutiveFailures = 0
		backoff = initialBackoff

		slog.Info("Change stream opened - waiting for documents", "stream", w.name)

		// Process the stream (inner loop)
		streamErr := w.processStream(stream)
		stream.Close(w.ctx)

		// Check why we exited
		if w.ctx.Err() != nil {
			// Context cancelled - clean shutdown
			return
		}

		if streamErr != nil {
			consecutiveFailures++
			slog.Warn("Change stream error, reconnecting",
				"error", streamErr,
				"stream", w.name,
				"attempt", consecutiveFailures,
				"backoff", backoff)

			// Check for stale resume token error
			if isStaleResumeTokenError(streamErr) {
				slog.Error("Resume token expired - clearing checkpoint and starting from current position. EVENTS MAY BE MISSED.",
					"stream", w.name)
				if w.checkpointStore != nil {
					w.clearCheckpoint()
				}
				backoff = initialBackoff // Don't backoff for stale token
				continue
			}

			// Wait before retry
			select {
			case <-w.ctx.Done():
				return
			case <-time.After(backoff):
			}

			// Increase backoff for next attempt
			backoff = time.Duration(float64(backoff) * backoffMultiplier)
			if backoff > maxBackoff {
				backoff = maxBackoff
			}
		}
	}
}

// processStream processes events from the change stream until an error or context cancellation
func (w *Watcher) processStream(stream *mongo.ChangeStream) error {
	batch := make([]bson.M, 0, w.config.BatchMaxSize)
	var lastToken bson.Raw
	batchStartTime := time.Now()

	for {
		select {
		case <-w.ctx.Done():
			// Flush remaining batch before exit
			if len(batch) > 0 {
				w.processBatch(batch, lastToken)
			}
			return nil

		default:
			// Check for next event with timeout
			ctx, cancel := context.WithTimeout(w.ctx, 100*time.Millisecond)
			hasNext := stream.TryNext(ctx)
			cancel()

			// Check for stream errors
			if err := stream.Err(); err != nil {
				// Flush batch before returning error
				if len(batch) > 0 {
					w.processBatch(batch, lastToken)
				}
				return err
			}

			if hasNext {
				var event bson.M
				if err := stream.Decode(&event); err != nil {
					slog.Error("Failed to decode change event", "error", err, "stream", w.name)
					continue
				}

				// Extract full document
				if fullDoc, ok := event["fullDocument"].(bson.M); ok {
					batch = append(batch, fullDoc)
					lastToken = stream.ResumeToken()
				}
			}

			// Check if we should flush the batch
			batchFull := len(batch) >= w.config.BatchMaxSize
			timeoutReached := time.Since(batchStartTime) >= w.config.BatchMaxWait

			if len(batch) > 0 && (batchFull || timeoutReached) {
				w.processBatch(batch, lastToken)
				batch = make([]bson.M, 0, w.config.BatchMaxSize)
				batchStartTime = time.Now()
			}
		}
	}
}

// isStaleResumeTokenError checks if the error is due to a stale/expired resume token
func isStaleResumeTokenError(err error) bool {
	if err == nil {
		return false
	}
	errStr := err.Error()
	// MongoDB error code 286 is ChangeStreamHistoryLost
	return contains(errStr, "ChangeStreamHistoryLost") ||
		contains(errStr, "resume token") ||
		contains(errStr, "oplog") ||
		contains(errStr, "invalidate")
}

// contains checks if s contains substr
func contains(s, substr string) bool {
	return len(s) >= len(substr) && (s == substr || (len(s) > 0 && findSubstring(s, substr)))
}

func findSubstring(s, substr string) bool {
	for i := 0; i <= len(s)-len(substr); i++ {
		if s[i:i+len(substr)] == substr {
			return true
		}
	}
	return false
}

// clearCheckpoint clears the checkpoint (for stale resume token recovery)
func (w *Watcher) clearCheckpoint() {
	if store, ok := w.checkpointStore.(*MongoCheckpointStore); ok {
		ctx, cancel := context.WithTimeout(context.Background(), 10*time.Second)
		defer cancel()
		_, err := store.collection.DeleteOne(ctx, bson.M{"_id": w.config.CheckpointKey})
		if err != nil {
			slog.Warn("Failed to clear checkpoint", "error", err, "stream", w.name)
		} else {
			slog.Info("Checkpoint cleared", "stream", w.name)
		}
	}
}

// processBatch processes a batch of documents
func (w *Watcher) processBatch(batch []bson.M, resumeToken bson.Raw) {
	if len(batch) == 0 {
		return
	}

	batchStartTime := time.Now()

	// Track batch metrics (like Java's StreamContext)
	currentSeq := w.batchSequence.Add(1)
	w.inFlightCount.Add(1)
	w.availableSlots.Add(-1)
	defer func() {
		w.inFlightCount.Add(-1)
		w.availableSlots.Add(1)
	}()

	slog.Debug("Processing batch",
		"stream", w.name,
		"batchSize", len(batch),
		"batchSeq", currentSeq)

	// Track success/failure counts for metrics
	successCount := 0
	failureCount := 0

	// Map and upsert documents
	for _, doc := range batch {
		docStartTime := time.Now()

		projected := w.mapper.Map(doc)
		if projected == nil {
			continue
		}

		// Get ID for upsert
		id, ok := projected["_id"]
		if !ok {
			slog.Warn("Projected document has no _id", "stream", w.name)
			failureCount++
			metrics.StreamEventsProcessed.WithLabelValues(w.name, "failed").Inc()
			continue
		}

		// Upsert to target collection
		filter := bson.M{"_id": id}
		update := bson.M{"$set": projected}
		opts := options.Update().SetUpsert(true)

		_, err := w.targetCollection.UpdateOne(w.ctx, filter, update, opts)
		if err != nil {
			slog.Error("Failed to upsert projection",
				"error", err,
				"stream", w.name,
				"id", id)
			failureCount++
			metrics.StreamEventsProcessed.WithLabelValues(w.name, "failed").Inc()
		} else {
			successCount++
			metrics.StreamEventsProcessed.WithLabelValues(w.name, "success").Inc()
		}

		// Record per-document processing duration
		metrics.StreamProcessingDuration.WithLabelValues(w.name).Observe(time.Since(docStartTime).Seconds())
	}

	// Save checkpoint and update checkpointed sequence
	if w.checkpointStore != nil && resumeToken != nil {
		if err := w.checkpointStore.SaveCheckpoint(w.config.CheckpointKey, resumeToken); err != nil {
			slog.Error("Failed to save checkpoint", "error", err, "stream", w.name)
		} else {
			w.checkpointedSeq.Store(currentSeq)
		}
	}

	// Update stream lag metric (difference between current and checkpointed sequence)
	lag := currentSeq - w.checkpointedSeq.Load()
	metrics.StreamLag.WithLabelValues(w.name).Set(float64(lag))

	slog.Debug("Batch processed",
		"stream", w.name,
		"processed", len(batch),
		"success", successCount,
		"failed", failureCount,
		"batchSeq", currentSeq,
		"duration", time.Since(batchStartTime))
}

// MongoCheckpointStore stores checkpoints in MongoDB
type MongoCheckpointStore struct {
	collection *mongo.Collection
}

// NewMongoCheckpointStore creates a new MongoDB checkpoint store
func NewMongoCheckpointStore(db *mongo.Database) *MongoCheckpointStore {
	return &MongoCheckpointStore{
		collection: db.Collection("stream_checkpoints"),
	}
}

// GetCheckpoint retrieves a checkpoint (resume token)
func (s *MongoCheckpointStore) GetCheckpoint(key string) (bson.Raw, error) {
	var doc struct {
		ResumeToken bson.Raw `bson:"resumeToken"`
	}

	err := s.collection.FindOne(context.Background(), bson.M{"_id": key}).Decode(&doc)
	if err != nil {
		if err == mongo.ErrNoDocuments {
			return nil, nil
		}
		return nil, err
	}

	// Return nil if no token stored (empty bson.Raw)
	if len(doc.ResumeToken) == 0 {
		return nil, nil
	}

	return doc.ResumeToken, nil
}

// SaveCheckpoint saves a checkpoint
func (s *MongoCheckpointStore) SaveCheckpoint(key string, token bson.Raw) error {
	filter := bson.M{"_id": key}
	update := bson.M{
		"$set": bson.M{
			"resumeToken": token,
			"updatedAt":   time.Now(),
		},
	}
	opts := options.Update().SetUpsert(true)

	_, err := s.collection.UpdateOne(context.Background(), filter, update, opts)
	return err
}

File: flowcatalyst-rust/crates/fc-stream/src/watcher.rs
//! Stream Watcher
//!
//! Watches MongoDB change streams and dispatches batches for concurrent processing.
//! Ensures aggregate ID ordering by not allowing the same aggregate in concurrent batches.

use crate::{StreamWatcher, StreamConfig};
use crate::checkpoint::CheckpointStore;
use crate::checkpoint_tracker::{CheckpointTracker, AggregateTracker, PendingDocument};
use async_trait::async_trait;
use mongodb::{Client, Collection};
use mongodb::bson::{doc, Document};
use mongodb::options::ChangeStreamOptions;
use mongodb::change_stream::event::ResumeToken;
use futures::stream::StreamExt;
use anyhow::Result;
use tracing::{info, warn, error, debug};
use std::sync::Arc;
use std::collections::HashSet;
use tokio::sync::Semaphore;
use tokio::time::{timeout, Duration};

/// Reconnection settings
const INITIAL_BACKOFF_MS: u64 = 5000;    // 5 seconds
const MAX_BACKOFF_MS: u64 = 60000;       // 60 seconds
const BACKOFF_MULTIPLIER: f64 = 2.0;

/// Batch processor callback
#[async_trait]
pub trait BatchProcessor: Send + Sync {
    /// Process a batch of documents
    async fn process(&self, documents: Vec<Document>) -> Result<()>;
}

/// Default batch processor that just logs
pub struct LoggingBatchProcessor {
    stream_name: String,
}

impl LoggingBatchProcessor {
    pub fn new(stream_name: &str) -> Self {
        Self {
            stream_name: stream_name.to_string(),
        }
    }
}

#[async_trait]
impl BatchProcessor for LoggingBatchProcessor {
    async fn process(&self, documents: Vec<Document>) -> Result<()> {
        info!("[{}] Processing batch of {} documents", self.stream_name, documents.len());
        Ok(())
    }
}

pub struct MongoStreamWatcher {
    client: Client,
    config: StreamConfig,
    checkpoint_store: Arc<dyn CheckpointStore>,
    checkpoint_tracker: Arc<CheckpointTracker>,
    aggregate_tracker: Arc<AggregateTracker>,
    batch_processor: Arc<dyn BatchProcessor>,
    concurrency_semaphore: Arc<Semaphore>,
}

impl MongoStreamWatcher {
    pub fn new(
        client: Client,
        config: StreamConfig,
        checkpoint_store: Arc<dyn CheckpointStore>,
        batch_processor: Arc<dyn BatchProcessor>,
    ) -> Self {
        let checkpoint_key = format!("checkpoint:{}", config.name);
        let checkpoint_tracker = Arc::new(CheckpointTracker::new(
            checkpoint_store.clone(),
            config.name.clone(),
            checkpoint_key,
        ));
        let aggregate_tracker = Arc::new(AggregateTracker::new(config.name.clone()));
        let concurrency_semaphore = Arc::new(Semaphore::new(config.concurrency as usize));

        Self {
            client,
            config,
            checkpoint_store,
            checkpoint_tracker,
            aggregate_tracker,
            batch_processor,
            concurrency_semaphore,
        }
    }

    /// Create with custom trackers (for testing or advanced usage)
    pub fn with_trackers(
        client: Client,
        config: StreamConfig,
        checkpoint_store: Arc<dyn CheckpointStore>,
        checkpoint_tracker: Arc<CheckpointTracker>,
        aggregate_tracker: Arc<AggregateTracker>,
        batch_processor: Arc<dyn BatchProcessor>,
    ) -> Self {
        let concurrency_semaphore = Arc::new(Semaphore::new(config.concurrency as usize));

        Self {
            client,
            config,
            checkpoint_store,
            checkpoint_tracker,
            aggregate_tracker,
            batch_processor,
            concurrency_semaphore,
        }
    }

    /// Check if the watcher should shutdown due to fatal error
    pub async fn has_fatal_error(&self) -> bool {
        self.checkpoint_tracker.has_fatal_error().await
    }

    /// Get the fatal error if one occurred
    pub async fn get_fatal_error(&self) -> Option<String> {
        self.checkpoint_tracker.get_fatal_error().await
    }

    /// Dispatch a batch for processing
    async fn dispatch_batch(
        &self,
        documents: Vec<Document>,
        aggregate_ids: HashSet<String>,
        resume_token: Option<ResumeToken>,
    ) -> Result<()> {
        if documents.is_empty() {
            return Ok(());
        }

        // Check for fatal error before dispatching
        if self.checkpoint_tracker.has_fatal_error().await {
            warn!("[{}] Skipping batch dispatch - fatal error occurred", self.config.name);
            return Ok(());
        }

        // Get batch sequence
        let seq = self.checkpoint_tracker.next_sequence().await;

        // Register aggregate IDs for this batch
        self.aggregate_tracker.register_batch(seq, aggregate_ids.clone()).await;

        // Acquire concurrency permit - blocks if at max concurrency
        let permit = self.concurrency_semaphore.clone().acquire_owned().await
            .map_err(|e| anyhow::anyhow!("Semaphore closed: {}", e))?;

        // Clone what we need for the spawned task
        let checkpoint_tracker = self.checkpoint_tracker.clone();
        let aggregate_tracker = self.aggregate_tracker.clone();
        let batch_processor = self.batch_processor.clone();
        let stream_name = self.config.name.clone();

        // Spawn task to process the batch
        tokio::spawn(async move {
            let _permit = permit; // Holds permit until task completes

            let result = batch_processor.process(documents).await;

            match result {
                Ok(()) => {
                    checkpoint_tracker.mark_complete(seq, resume_token).await;
                    debug!("[{}] Batch {} completed", stream_name, seq);
                }
                Err(e) => {
                    error!("[{}] Batch {} failed: {}", stream_name, seq, e);
                    checkpoint_tracker.mark_failed(seq, e.to_string()).await;
                }
            }

            // Release aggregate IDs and check for pending documents
            let released = aggregate_tracker.complete_batch(seq).await;
            if !released.is_empty() {
                debug!("[{}] Released {} pending documents after batch {}", stream_name, released.len(), seq);
            }
        });

        Ok(())
    }

    /// Extract aggregate ID from a document
    fn get_aggregate_id(&self, doc: &Document) -> Option<String> {
        // Try to get the aggregate ID field
        let field = &self.config.aggregate_id_field;

        if let Ok(id) = doc.get_str(field) {
            return Some(id.to_string());
        }

        // Try as ObjectId
        if let Ok(id) = doc.get_object_id(field) {
            return Some(id.to_hex());
        }

        // Try nested path (e.g., "_id")
        if field == "_id" {
            if let Some(bson) = doc.get("_id") {
                return Some(format!("{}", bson));
            }
        }

        None
    }
}

#[async_trait]
impl StreamWatcher for MongoStreamWatcher {
    async fn watch(&self) -> Result<()> {
        let db = self.client.database(&self.config.source_database);
        let collection: Collection<Document> = db.collection(&self.config.source_collection);
        let checkpoint_key = format!("checkpoint:{}", self.config.name);

        let mut consecutive_failures = 0u32;
        let mut backoff_ms = INITIAL_BACKOFF_MS;

        // Outer reconnection loop
        loop {
            // Check for fatal error
            if self.checkpoint_tracker.has_fatal_error().await {
                error!("[{}] Fatal error detected - stopping watcher", self.config.name);
                return Err(anyhow::anyhow!("Fatal error in batch processing"));
            }

            // Load checkpoint
            let resume_token_doc = match self.checkpoint_store.get_checkpoint(&checkpoint_key).await {
                Ok(doc) => doc,
                Err(e) => {
                    warn!("[{}] Failed to load checkpoint, starting from current: {}", self.config.name, e);
                    None
                }
            };

            let mut options = ChangeStreamOptions::builder()
                .full_document(Some(mongodb::options::FullDocumentType::UpdateLookup))
                .build();

            if let Some(doc) = resume_token_doc {
                info!("[{}] Resuming from checkpoint", self.config.name);
                if let Ok(token) = mongodb::bson::from_document::<ResumeToken>(doc) {
                    options.resume_after = Some(token);
                }
            } else {
                info!("[{}] Starting from current position (no checkpoint)", self.config.name);
            }

            let pipeline = vec![
                doc! { "$match": { "operationType": { "$in": &self.config.watch_operations } } }
            ];

            // Try to open the change stream
            let stream_result = collection.watch().pipeline(pipeline).with_options(options).await;
            let mut stream = match stream_result {
                Ok(s) => {
                    consecutive_failures = 0;
                    backoff_ms = INITIAL_BACKOFF_MS;
                    info!("[{}] Change stream opened on {}.{} (concurrency: {})",
                        self.config.name, self.config.source_database, self.config.source_collection,
                        self.config.concurrency);
                    s
                }
                Err(e) => {
                    consecutive_failures += 1;

                    if is_stale_resume_token_error(&e) {
                        error!("[{}] Resume token expired - clearing checkpoint. EVENTS MAY BE MISSED.", self.config.name);
                        let _ = self.checkpoint_store.clear_checkpoint(&checkpoint_key).await;
                        backoff_ms = INITIAL_BACKOFF_MS;
                        continue;
                    }

                    error!("[{}] Failed to open change stream (attempt {}), retrying in {}ms: {}",
                        self.config.name, consecutive_failures, backoff_ms, e);

                    tokio::time::sleep(Duration::from_millis(backoff_ms)).await;
                    backoff_ms = (backoff_ms as f64 * BACKOFF_MULTIPLIER) as u64;
                    if backoff_ms > MAX_BACKOFF_MS {
                        backoff_ms = MAX_BACKOFF_MS;
                    }
                    continue;
                }
            };

            // Process stream events
            let stream_error = self.process_stream_events(&mut stream).await;

            match stream_error {
                Ok(()) => {
                    info!("[{}] Change stream ended cleanly", self.config.name);
                    return Ok(());
                }
                Err(e) => {
                    // Check if this is a fatal error from batch processing
                    if self.checkpoint_tracker.has_fatal_error().await {
                        error!("[{}] Fatal error in batch processing - stopping", self.config.name);
                        return Err(e);
                    }

                    consecutive_failures += 1;

                    if is_stale_resume_token_error(&e) {
                        error!("[{}] Resume token expired - clearing checkpoint. EVENTS MAY BE MISSED.", self.config.name);
                        let _ = self.checkpoint_store.clear_checkpoint(&checkpoint_key).await;
                        backoff_ms = INITIAL_BACKOFF_MS;
                        continue;
                    }

                    warn!("[{}] Change stream error (attempt {}), reconnecting in {}ms: {}",
                        self.config.name, consecutive_failures, backoff_ms, e);

                    tokio::time::sleep(Duration::from_millis(backoff_ms)).await;
                    backoff_ms = (backoff_ms as f64 * BACKOFF_MULTIPLIER) as u64;
                    if backoff_ms > MAX_BACKOFF_MS {
                        backoff_ms = MAX_BACKOFF_MS;
                    }
                }
            }
        }
    }
}

impl MongoStreamWatcher {
    /// Process events from an active change stream
    async fn process_stream_events(
        &self,
        stream: &mut mongodb::change_stream::ChangeStream<mongodb::change_stream::event::ChangeStreamEvent<Document>>,
    ) -> Result<()> {
        let mut batch: Vec<Document> = Vec::new();
        let mut batch_aggregate_ids: HashSet<String> = HashSet::new();
        let mut last_token: Option<ResumeToken> = None;
        let batch_timeout = Duration::from_millis(self.config.batch_max_wait_ms);

        loop {
            // Check for fatal error
            if self.checkpoint_tracker.has_fatal_error().await {
                error!("[{}] Fatal error detected - stopping stream processing", self.config.name);
                // Flush current batch before returning
                if !batch.is_empty() {
                    let _ = self.dispatch_batch(
                        std::mem::take(&mut batch),
                        std::mem::take(&mut batch_aggregate_ids),
                        last_token.clone(),
                    ).await;
                }
                return Err(anyhow::anyhow!("Fatal error in batch processing"));
            }

            let event_result = timeout(batch_timeout, stream.next()).await;

            match event_result {
                Ok(Some(Ok(event))) => {
                    if let Some(doc) = event.full_document {
                        // Get aggregate ID
                        let aggregate_id = self.get_aggregate_id(&doc);

                        if let Some(ref agg_id) = aggregate_id {
                            // Check if this aggregate is already in-flight
                            if self.aggregate_tracker.is_in_flight(agg_id).await {
                                // Add to pending queue
                                debug!("[{}] Aggregate {} is in-flight, queueing document", self.config.name, agg_id);
                                self.aggregate_tracker.add_pending(PendingDocument {
                                    aggregate_id: agg_id.clone(),
                                    document: doc,
                                    resume_token: stream.resume_token(),
                                }).await;
                                continue;
                            }

                            // Check if this aggregate is already in current batch
                            if batch_aggregate_ids.contains(agg_id) {
                                // Flush current batch first, then add this to next batch
                                if !batch.is_empty() {
                                    self.dispatch_batch(
                                        std::mem::take(&mut batch),
                                        std::mem::take(&mut batch_aggregate_ids),
                                        last_token.clone(),
                                    ).await?;
                                }
                            }

                            batch_aggregate_ids.insert(agg_id.clone());
                        }

                        batch.push(doc);
                        last_token = stream.resume_token();
                    }

                    // Check if batch is full
                    if batch.len() >= self.config.batch_max_size as usize {
                        self.dispatch_batch(
                            std::mem::take(&mut batch),
                            std::mem::take(&mut batch_aggregate_ids),
                            last_token.clone(),
                        ).await?;
                    }
                }
                Ok(Some(Err(e))) => {
                    // Stream error - flush batch and return for reconnection
                    if !batch.is_empty() {
                        let _ = self.dispatch_batch(
                            std::mem::take(&mut batch),
                            std::mem::take(&mut batch_aggregate_ids),
                            last_token.clone(),
                        ).await;
                    }
                    return Err(e.into());
                }
                Ok(None) => {
                    // Stream closed
                    if !batch.is_empty() {
                        let _ = self.dispatch_batch(
                            std::mem::take(&mut batch),
                            std::mem::take(&mut batch_aggregate_ids),
                            last_token.clone(),
                        ).await;
                    }
                    return Err(anyhow::anyhow!("Change stream closed unexpectedly"));
                }
                Err(_) => {
                    // Timeout - flush batch if any
                    if !batch.is_empty() {
                        self.dispatch_batch(
                            std::mem::take(&mut batch),
                            std::mem::take(&mut batch_aggregate_ids),
                            last_token.clone(),
                        ).await?;
                    }
                }
            }
        }
    }
}

/// Check if an error is due to a stale/expired resume token
fn is_stale_resume_token_error<E: std::fmt::Display>(e: &E) -> bool {
    let err_str = e.to_string().to_lowercase();
    err_str.contains("changestream") && err_str.contains("history") ||
    err_str.contains("resume token") ||
    err_str.contains("oplog") ||
    err_str.contains("invalidate")
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_stale_resume_token_detection() {
        assert!(is_stale_resume_token_error(&"ChangeStream history lost"));
        assert!(is_stale_resume_token_error(&"resume token expired"));
        assert!(is_stale_resume_token_error(&"oplog entry not found"));
        assert!(!is_stale_resume_token_error(&"connection refused"));
    }
}
